packages/backend/server/src/plugins/copilot/context/resolver.ts

```diff
@@ -0,0 +1,446 @@
+    return await session.listFiles();
+  }
+
+  @Mutation(() => String, {
+    description: 'add a file to context',
+  })
+  @CallMetric('ai', 'context_file_add')
+  async addContextFile(
+    @Context() ctx: { req: Request },
+    @Args({ name: 'options', type: () => AddContextFileInput })
+    options: AddContextFileInput,
+    @Args({ name: 'content', type: () => GraphQLUpload })
+    content: FileUpload
+  ) {
+    const lockFlag = `${COPILOT_LOCKER}:context:${options.contextId}`;
+    await using lock = await this.mutex.acquire(lockFlag);
+    if (!lock) {
+      return new TooManyRequest('Server is busy');
+    }
+    const session = await this.context.get(options.contextId);
+
+    try {
+      const signal = this.getSignal(ctx.req);
+      return await session.addStream(
+        content.createReadStream(),
+        content.filename,
+        options.blobId,
+        signal
+      );
+    } catch (e: any) {
+      throw new CopilotFailedToModifyContext({
+        contextId: options.contextId,
+        message: e.message,
+      });
+    }
+  }
+
+  @Mutation(() => Boolean, {
+    description: 'remove a file from context',
+  })
+  @CallMetric('ai', 'context_file_remove')
+  async removeContextFile(
+    @Args({ name: 'options', type: () => RemoveContextFileInput })
+    options: RemoveContextFileInput
+  ) {
+    const lockFlag = `${COPILOT_LOCKER}:context:${options.contextId}`;
+    await using lock = await this.mutex.acquire(lockFlag);
+    if (!lock) {
+      return new TooManyRequest('Server is busy');
+    }
+    const session = await this.context.get(options.contextId);
+
+    try {
+      return await session.remove(options.fileId);
+    } catch (e: any) {
+      throw new CopilotFailedToModifyContext({
+        contextId: options.contextId,
+        message: e.message,
+      });
+    }
+  }
+
+  @Mutation(() => [ContextMatchedFileChunk], {
+    description: 'remove a file from context',
+  })
+  @CallMetric('ai', 'context_file_remove')
+  async matchContext(
+    @Context() ctx: { req: Request },
+    @Args('contextId') contextId: string,
+    @Args('content') content: string,
+    @Args('limit', { type: () => SafeIntResolver, nullable: true })
+    limit?: number
+  ) {
+    const lockFlag = `${COPILOT_LOCKER}:context:${contextId}`;
+    await using lock = await this.mutex.acquire(lockFlag);
+    if (!lock) {
+      return new TooManyRequest('Server is busy');
+    }
+    const session = await this.context.get(contextId);
+
+    try {
+      return await session.match(content, limit, this.getSignal(ctx.req));
+    } catch (e: any) {
+      throw new CopilotFailedToMatchContext({
+        contextId,
+        // don't record the large content
+        content: content.slice(0, 512),
+        message: e.message,
+      });
+    }
+  }
+
+  @Mutation(() => ContextMatchedDocChunk, {
+    description: 'match workspace doc',
+  })
+  @CallMetric('ai', 'context_match_workspace_doc')
+  async matchWorkspaceContext(
```

Comment:
We will provide two ways of document embedding in the future: full document embedding and the blocks embedding. I think we should differentiate the APIs in this dimension, not workspace latitude.

Reply:
even if block-level indexing is supported in the future, the search will still be conducted within the workspace, only requiring differentiation through parameters
secondly, there is no scheduled block-level reading for the server in the short term